### ISSUE TITLE
Cannot intersect a polygon with a line

### ISSUE THREAD

**Issue by zhaishengfu** [Xrayez/godot-clipper#2]
ahother question. when i intersect using a line and a polygon, the result is not correct：
	
	hori = PoolVector2Array( Array([Vector2(0, 600), Vector2( texturedata.get_width(), 0)] ) ) 
	clipper.set_mode(Clipper.MODE_CLIP)
	clipper.path_type = Clipper.ptSubject
	clipper.add_points(c[0])
	clipper.path_type = Clipper.ptClip
	clipper.add_points(hori)
	clipper.clip_type = Clipper.ctIntersection
	clipper.execute()
i am sure the line intersect with the polygon of c[0], can it use open line to intersect?

**Comment by Xrayez** [comment]
No, unfortunately Clipper doesn't support this, open paths (lines) can only be subject...

You also forgot to specify it's actually open (wouldn't work either way)
```gdscript
clipper.open = true
clipper.add_points(hori)
```

I should add the default parameter for this I think: `clipper.add_points(hori, true)`

This surely can be confusing at first but this is how I managed to combine the three Clipper units together having the same interface. Me and original author are currently working on finishing Clipper 10.0.0, after it's declared stable I'll continue improving the module.

Also see godotengine/godot#23559, so this module might be merged at some point.

**Comment by zhaishengfu** [comment]
Sorry, but you said" Clipper doesn't support this", what is your meaning? i see official clipperLib supports this, http://www.angusj.com/delphi/clipper/documentation/Docs/Units/ClipperLib/Types/ClipType.htm

PS: i tried set open, but the result is also wrong

**Comment by zhaishengfu** [comment]
and can you help me do the following effect? in an image, i want to add multi horizontal lines and vertical lines to intersect it and get all the interact points between the image and all lines. i am thinking using box instead of lines, but this seems inconvenient
![image](https://user-images.githubusercontent.com/7630101/54806731-e10c9300-4cb5-11e9-8afd-eb7c66f31204.png)

i want to get the following effect:
![image](https://user-images.githubusercontent.com/7630101/54806779-08636000-4cb6-11e9-99fe-e5b350e766e9.png)

**Comment by zhaishengfu** [comment]
and ， can you offer a document of your api and some examples? for example, 
intersect( PoolVector2Array poly_a, PoolVector2Array poly_b, bool is_a_open=false )

how to use this function? and the third variable is open, how to use it ?

**Comment by zhaishengfu** [comment]
i think up a new method， can your library judge whether a point is in a polygon?? if so, i can use this instead of intersect line with polygon

**Comment by Xrayez** [comment]
> Sorry, but you said" Clipper doesn't support this", what is your meaning?

It works only for:
`clipper.intersect(polyline_subject, polygon_clip)`
but not:
`clipper.intersect(polygon_subject, polyline_clip)`

Which in your example is what you're trying to do.

> i am thinking using box instead of lines, but this seems inconvenient

Yes this is one of the ways to do this. I think I misunderstood your intention, so this should be possible to do so like this:
```gdscript
hori = PoolVector2Array( Array([Vector2(0, 600), Vector2( texturedata.get_width(), 0)] ) ) 
clipper.set_mode(Clipper.MODE_CLIP)
# Lines can only be subject
clipper.path_type = Clipper.ptSubject # changed to subject
clipper.add_points(hori)
clipper.path_type = Clipper.ptClip # changed to clip
clipper.add_points(c[0])
clipper.clip_type = Clipper.ctIntersection
clipper.execute()
```

> can your library judge whether a point is in a polygon?

No, it's not present in this version for some reason, but was present in the previous version 6.4.2. `PointInPolygon` method is present in the most recent version 10.0.0 which isn't ported to C++ completely yet.

> can you offer a document of your api and some examples?

Sorry I won't be working on documentation until stable version comes out. I'm not the one who wrote the original library either, so most of the problems you're describing pertains to library itself, so you could refer to Clipper's documentation for now.

**Comment by zhaishengfu** [comment]
ok .thank you for your reply. i got the detailed information. but your method

hori = PoolVector2Array( Array([Vector2(0, 600), Vector2( texturedata.get_width(), 0)] ) ) 
clipper.set_mode(Clipper.MODE_CLIP)

clipper.path_type = Clipper.ptSubject # changed to subject
clipper.add_points(hori)
clipper.path_type = Clipper.ptClip # changed to clip
clipper.add_points(c[0])
clipper.clip_type = Clipper.ctIntersection
clipper.execute()

does not work..

the solution count is zero

**Comment by Xrayez** [comment]
The closed and open paths solutions are separated in solution, you should use:
```gdscript
for idx in get_solution_count(Clipper.TYPE_OPEN):
    var line = clipper.get_solution(idx, Clipper.TYPE_OPEN)
```
You could also avoid all this by:
```gdscript
var is_hori_open = true
var lines = clipper.intersect(hori, c[0], is_hori_open)
```

**Comment by zhaishengfu** [comment]
hello， i tried both of your method. the second method works!. but for first method, there are problems. for fear of misunderstanding, i will paste my full code and add node. 
for your Clipper.TYPE_OPEN method, i tried the following code:
(1)  **without** **clipper.open** and using **Clipper.TYPE_OPEN** and **ctUnion** 

```
func _ready():
        hori = PoolVector2Array( Array([Vector2(0, 600), Vector2( texturedata.get_width()+300, 0)] ) )  # an open line
	clipper.set_mode(Clipper.MODE_CLIP)
	clipper.path_type = Clipper.ptSubject
	# clipper.open = true   # delete this for test
	clipper.add_points(hori)
	clipper.path_type = Clipper.ptClip
	clipper.add_points(bb) #bb is my image contour 
	clipper.clip_type = Clipper.ctUnion
	clipper.execute()
        ...

func _process():
        for idx in clipper.get_solution_count(0):
		print(idx)
		var points = clipper.get_solution(idx, 0)
		for i in range(0, points.size()-1):
			draw_line(points[i], points[i+1], Color(0.5*idx, 1, 1, 1))
		draw_line(points[points.size()-1], points[0], Color(0, 0, 1, 1))
```

the result is following(all the lines, different color is only for seeing and does not matter)
![image](https://user-images.githubusercontent.com/7630101/54860734-45346300-4d59-11e9-8246-9b07bfd24c69.png)
(2) i change the (1) code using **TYPE_OPEN**, the result is none
![image](https://user-images.githubusercontent.com/7630101/54860745-8593e100-4d59-11e9-8530-b022aad9c6e0.png)
(3) i change the (1) code using **clipper_type = Clipper.ctIntersection**, the result is none
(4) i change the (1) code by **adding clipper.open = true**, and **clipe_type = Clipper.ctIntersection** and using **TYPE_OPEN**, the result is none
(5) i change the (1) code by **adding clipper.open = true, and clipe_type = Clipper.ctUnion and using TYPE_OPEN**, the result is following line
![image](https://user-images.githubusercontent.com/7630101/54860793-3c905c80-4d5a-11e9-93d2-898df8d069b9.png)

**so to sum up：**
（1） whether use TYPE_OPEN or TYPE_CLOSE, only ctUnion works
  (2)    clipper.open = true only works when the clip_type =  ctUnion
i thinks this is really confusing. i thinks ctInteraction should also work

**Comment by Xrayez** [comment]
Yep, forgot to reset `open` parameter for image contour:
```gdscript
func _ready():
        hori = PoolVector2Array( Array([Vector2(0, 600), Vector2( texturedata.get_width()+300, 0)] ) )
	clipper.set_mode(Clipper.MODE_CLIP)

	clipper.path_type = Clipper.ptSubject
	clipper.open = true # line
	clipper.add_points(hori)

	clipper.path_type = Clipper.ptClip
        # Have to reset `open` property so that the next added path 
        # will be treated as closed polygon
        clipper.open = false # polygon
	clipper.add_points(bb) #bb is my image contour 

	clipper.clip_type = Clipper.ctIntersection
	clipper.execute()
```

I see how this can be confusing and will take this into consideration when rewriting the module.

Also, depending on how you extract the contour (clockwise/anticlockwise) might also interfere with clipper interpreting those as holes or boundary polygons respectively, but not sure about this because it might fix polygon orientation for you internally.

I've also uploaded an example project for testing open paths:
https://github.com/Xrayez/godot-clipper/tree/master/examples/projects/open_paths

**Comment by zhaishengfu** [comment]
ops, I get it. thanks

**Comment by Xrayez** [comment]
Closing in favor of [GeometryTools](https://github.com/Xrayez/godot-geomtools) module.

> i think up a new method， can your library judge whether a point is in a polygon?? if so, i can use this instead of intersect line with polygon

Possible now in the mentioned module with `GeometryTools2D.point_in_polygon()` 🙂

This module is also deprecated: dbf602a.
